core/devdoc/on-out-process-gateway-modules.md

```diff
@@ -0,0 +1,484 @@
+using the ipc transport provided in *nanomsg*. This will be the default 
+transport for outprocess modules when a URI is not given. *Nanomsg* supports TCP,
+but it is currently not secured and therefore not recommended until a secure transport is available.
+
+Proxy Module
+------------
+
+The first piece of the puzzle is a gateway module (i.e., it implements
+`MODULE_API`) that we shall call the *proxy* module. The job of the proxy 
+module is to abstract from the rest of the gateway the fact that the module in 
+question is running outside the gateway process, i.e., as far as the rest of 
+the gateway is concerned, the out-proc module is no different from the garden 
+variety in-proc module. The proxy module’s out-proc counterpart is a *hosting 
+process* that is responsible for running the outprocess module in question and 
+broker communication between that module and the gateway.
+
+This is a brief description of the repsonsibilies of the proxy module.
+
+-   **Module_ParseConfigurationFromJson**
+    The proxy module will serialize the arguments into a JSON string.
+
+-   **Module_FreeConfiguration**
+    The proxy module will free the serialized string.
+
+-   **Module_Create**
+    The proxy Module will format a _Create Module_ request and send it via the 
+    control channel to the outprocess module, it will wait for a response 
+    before returning.
+
+-   **Module_Destroy**
+    The proxy Module will format a _Destroy Module_ message and send it via the control channel to the outprocess module. It will not wait for a response.
+
+-   **Module_Receive**
+    The proxy module will serialize the message and send it via the message channel to the outprocess module.
+
+-   **Module_Start**
+    The proxy module will format a _Start Module_ message and send it via the control channel to the outprocess module.
+
+-   The proxy module will also create a task to read any gateway messages 
+    published from the remote module.
+
+The proxy module will require a new loader type, an outprocess module loader.
+
+Outprocess Module loader
+-------------------
+
+The proxy module will be loaded by a new "outprocess" module loader type. This will be responsible for providing the proxy module API to the gateway.
+
+There is no configuration for the proxy module loader.
+
+The proxy module connection information is placed in the "entrypoint" for the loader.
+
+This is an example of what the "loaders" entry will look like this:
+```JSON
+"loaders" : [
+    { 
+        "type" : "outprocess",
+        "name" : "outprocess"
+    }
+]
+```
+
+and an example of how the proxy module will be loaded :
+
+```JSON
+"modules" : [
+{ 
+    "name" : "outprocess_module",
+    "loader" : {
+        "name" : "outprocess"
+        "entrypoint" : {
+            "activation.type" : "none",
+            "control.id" : "outprocess_module_control",
+            "message.id" : "outprocess_module_message",
+            }
```

Comment:
The message ID is optional.  If you specify one, we use it, otherwise, the loader creates a name.